src/targets/gpu/fuse_mlir.cpp

```diff
@@ -678,93 +680,140 @@ using find_mlir_standalone_dot_op         = find_mlir_standalone_op<&is_mlir_dot
+
     auto matcher() const
     {
-        return match::name("gpu::pre_gemm_softmax_gemm").bind("gemm_softmax_gemm");
+        auto gemm1 =
+            match::skip(match::name("contiguous"))(match::used_once(), is_mlir_dot(dot_mode))
+                .bind("gemm1");
+        auto fused_reduce =
+            match::name("fused_reduce")(match::used_once(), match::any_of[match::inputs()](gemm1))
+                .bind("fused_reduce");
+        return is_mlir_dot(dot_mode)(match::arg(0)(fused_reduce)).bind("gemm2");
     }
 
-    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    std::unordered_map<instruction_ref, instruction_ref>
+    invert_map_ins(const std::unordered_map<instruction_ref, instruction_ref>& map_ins) const
     {
-        static size_t counter  = 0;
-        module_ref mm          = mpm.create_module("mlir_" + std::to_string(counter++));
-        auto gemm_softmax_gemm = r.instructions["gemm_softmax_gemm"];
-        mm->set_bypass();
+        std::unordered_map<instruction_ref, instruction_ref> inverse_map;
+        for(auto const& [i1, i2] : map_ins)
+        {
+            assert(not contains(inverse_map, i2));
+            inverse_map[i2] = i1;
+        }
+        return inverse_map;
+    }
 
-        auto orig_inputs = gemm_softmax_gemm->inputs();
+    auto finalize_attention_module(module_ref m) const
+    {
+        eliminate_common_subexpression{}.apply(*m);
+        dead_code_elimination{}.apply(*m);
+    }
 
-        std::vector<instruction_ref> gemm0_inputs = {orig_inputs[0], orig_inputs[1]};
-        auto [gemm0, top_gemm0_inputs] =
-            fuse_input_ops_and_gemm_based_op(mm, gemm0_inputs, make_op("dot"));
+    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    {
+        auto gemm2        = r.instructions["gemm2"];
+        auto fused_reduce = r.instructions["fused_reduce"];
+        auto gemm1        = r.instructions["gemm1"];
 
-        std::vector<instruction_ref> inputs;
-        inputs.insert(inputs.begin(), top_gemm0_inputs.begin(), top_gemm0_inputs.end());
-
-        // handle scale
-        auto v = gemm_softmax_gemm->get_operator().to_value();
-        assert(v.contains("scale"));
-        auto scale     = v.at("scale").to<float>();
-        auto scale_lit = mm->add_literal(literal{shape{gemm0->get_shape().type()}, {scale}});
-        instruction_ref scale_lit_mbcast = mm->add_instruction(
-            make_op("multibroadcast", {{"out_lens", gemm0->get_shape().lens()}}), scale_lit);
-        auto scaled_gemm0 = mm->add_instruction(make_op("mul"), gemm0, scale_lit_mbcast);
-
-        std::optional<instruction_ref> bias{nullopt};
-        if(orig_inputs.size() == 4)
+        auto axes = fused_reduce->get_operator().to_value()["axes"];
+        if(axes.size() != 1)
+            return;
+
+        module_ref m_attn =
+            mpm.create_module("mlir_attn_" + fused_reduce->module_inputs().front()->name());
```

Comment:
Instead of creating the module and then removing if it doesn't work, you could just construct a module variable directly: `module m_attn` and then later if it can be added as attention you can do `mpm.create_module("mlir_attn_" + fused_reduce->module_inputs().front()->name(), std::move(m_attn))`.